relay-filter/src/config.rs

```diff
@@ -248,13 +251,177 @@ impl GenericFilterConfig {
+///                 id: "filter1",
+///                 is_enabled: false,
+///                 condition: None,
+///             },
+///         },
+///     }
+/// "#);
+/// ```
+///
+/// Deserialization of no filters defaults to an empty map:
+///
+/// ```
+/// # use relay_filter::GenericFiltersConfig;
+/// # use insta::assert_debug_snapshot;
+///
+/// let json = r#"{
+///     "version": 1
+/// }"#;
+/// let deserialized = serde_json::from_str::<GenericFiltersConfig>(json).unwrap();
+/// assert_debug_snapshot!(deserialized, @r#"
+///     GenericFiltersConfig {
+///         version: 1,
+///         filters: {},
+///     }
+/// "#);
+/// ```
+///
+/// Serialization:
+///
+/// ```
+/// # use relay_filter::{GenericFiltersConfig, GenericFilterConfig};
+/// # use relay_protocol::condition::RuleCondition;
+/// # use insta::assert_display_snapshot;
+/// # use indexmap::IndexMap;
+///
+/// let filter = GenericFiltersConfig {
+///     version: 1,
+///     filters: IndexMap::from([(
+///         "filter1".to_owned(),
+///         GenericFilterConfig {
+///             id: "filter1".to_owned(),
+///             is_enabled: true,
+///             condition: Some(RuleCondition::eq("event.exceptions", "drop-error")),
+///         },
+///     )]),
+/// };
+/// let serialized = serde_json::to_string_pretty(&filter).unwrap();
+/// assert_display_snapshot!(serialized, @r#"{
+///   "version": 1,
+///   "filters": [
+///     {
+///       "id": "filter1",
+///       "isEnabled": true,
+///       "condition": {
+///         "op": "eq",
+///         "name": "event.exceptions",
+///         "value": "drop-error"
+///       }
+///     }
+///   ]
+/// }"#);
+/// ```
+///
+/// Serialization of filters is skipped if there aren't any:
+///
+/// ```
+/// # use relay_filter::{GenericFiltersConfig, GenericFilterConfig};
+/// # use relay_protocol::condition::RuleCondition;
+/// # use insta::assert_display_snapshot;
+/// # use indexmap::IndexMap;
+///
+/// let filter = GenericFiltersConfig {
+///     version: 1,
+///     filters: IndexMap::new(),
+/// };
+/// let serialized = serde_json::to_string_pretty(&filter).unwrap();
+/// assert_display_snapshot!(serialized, @r#"{
+///   "version": 1
+/// }"#);
+/// ```
 #[derive(Clone, Debug, Default, Serialize, Deserialize)]
-pub(crate) struct GenericFiltersConfig {
+#[serde(rename_all = "camelCase")]
+pub struct GenericFiltersConfig {
     /// Version of the filters configuration.
     pub version: u16,
-    /// List of generic filters.
-    pub filters: Vec<GenericFilterConfig>,
+    /// Map of generic filters, sorted by the order in the payload from upstream.
+    ///
+    /// The map contains unique filters, meaning there are no two filters with
+    /// the same id. See struct docs for more details.
+    #[serde(
+        default,
+        skip_serializing_if = "IndexMap::is_empty",
+        with = "generic_filters_custom_serialization"
+    )]
+    pub filters: IndexMap<String, GenericFilterConfig>,
```

Comment:
I wonder if we should make this a newtype instead and hide the `IndexMap` which should make it impossible to construct an invalid index map. 

Can also implement Serialize/Deserialize directly on the newtype (cannot forget to add the annotation)

Reply:
Done in [d7790ea](https://github.com/getsentry/relay/pull/3161/commits/d7790eab2cb3ba91d9269a1bddb685314db7b632).